### PR TITLE
Create some apis and test cases for mapping asset name and ip

### PR DESCRIPTION
**Pull request by Pengpengwanga**

### CODE REVIEW

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 1]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
    @RequestMapping(value = "/mapping/hostnameip", method = RequestMethod.POST)
    public void createHostNameIPMapping(@RequestBody AssetIPMapping mapping) {
       BaseDocumentUtil.generateID(mapping);
+      String ip = mapping.getIp();
+      if(ip == null || ip.isEmpty() || !ip.matches(HandleURL.IP_REGX)) {
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:553]: Let's use the existing method to check ip.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 2]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+         throw new WormholeRequestException(HttpStatus.INTERNAL_SERVER_ERROR, "Invalid ip address",
+               null);
+      }
+      if (!assetService.isAssetNameValidate(mapping)) {
+         throw new WormholeRequestException(HttpStatus.INTERNAL_SERVER_ERROR, "Invalid asset name",
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:558]: let' change the error message more meanful.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 3]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+               null);
+      }
       assetIPMappingRepository.save(mapping);
    }
 
+   @ResponseStatus(HttpStatus.OK)
+   @RequestMapping(value = "/mapping/hostnameip", method = RequestMethod.PUT)
+   public void updateHostNameIPMapping(@RequestBody AssetIPMapping mapping) {
+      BaseDocumentUtil.generateID(mapping);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:567]: why?

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 4]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+      AssetIPMapping oldMapping = assetIPMappingRepository.findOne(mapping.getId());
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:568]: should check no mapping find case.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 5]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+      if(!assetService.isAssetNameValidate(mapping)) {
+         throw new WormholeRequestException(HttpStatus.INTERNAL_SERVER_ERROR,"Invalid asset name",null);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:570]: message correct.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 6]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+      }
+      oldMapping.setAssetname(mapping.getAssetname());
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:572]: check if name not changed then just return.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 7]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+      assetIPMappingRepository.save(oldMapping);
+   }
+
+   @ResponseStatus(HttpStatus.OK)
+   @RequestMapping(value = "/mapping/hostnameip", method = RequestMethod.GET)
+   public Page<AssetIPMapping> getHostNameByRange(@RequestParam(value = "pagesize",required = false, defaultValue = "20") Integer pageSize,
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:578]: ByPage

Comment (yixingjia) [reply flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:578]: Method name should change to getAssetIPMappingByPage.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 8]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+         @RequestParam(value = "pagenumber",required = false, defaultValue = "1") Integer pageNumber) {
+      if (pageNumber < FlowgateConstant.defaultPageNumber) {
+         pageNumber = FlowgateConstant.defaultPageNumber;
+      } else if (pageSize <= 0) {
+         pageSize = FlowgateConstant.defaultPageSize;
+      } else if (pageSize > FlowgateConstant.maxPageSize) {
+         pageSize = FlowgateConstant.maxPageSize;
+      }
+      PageRequest pageRequest = new PageRequest(pageNumber - 1, pageSize);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:587]: Let's create a job to search mapping by IP , then we should allow user to edit the IPMapping base on the search result.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 9]
```diff
@@ -548,9 +549,57 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+   }
+
+   @ResponseStatus(HttpStatus.OK)
+   @RequestMapping(value = "/mapping/hostnameip/{id}", method = RequestMethod.DELETE)
+   public void deleteHostNameByID(@PathVariable("id") String id) {
+      assetIPMappingRepository.delete(id);
+   }
+
+   @ResponseStatus(HttpStatus.OK)
+   @RequestMapping(value = "/name/keywords/{keywords}", method = RequestMethod.GET)
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:598]: /names?queryParam=xxx

File: flowgate-api/src/main/java/com/vmware/flowgate/util/HandleURL.java [hunk 10]
```diff
@@ -8,6 +8,11 @@
 
 public class HandleURL {
 
+   public static final String IP_REGX = "^(1\\d{2}|2[0-4]\\d|25[0-5]|[1-9]\\d|[1-9])\\."
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/util/HandleURL.java:11]: should be able to remove it.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 11]
```diff
@@ -548,9 +549,63 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
 
+   @ResponseStatus(HttpStatus.OK)
+   @RequestMapping(value = "/mapping/hostnameip", method = RequestMethod.PUT)
+   public void updateHostNameIPMapping(@RequestBody AssetIPMapping mapping) {
+      AssetIPMapping oldMapping = assetIPMappingRepository.findOne(mapping.getId());
+      if(oldMapping == null) {
+         throw new WormholeRequestException(HttpStatus.INTERNAL_SERVER_ERROR,"The dd is not exist : " + mapping.getId(),null);
+      }
+      String assetName = mapping.getAssetname();
+      if(assetName == null || !assetService.isAssetNameValidate(assetName)) {
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:572]: double check if we already check empty in the validate method.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 12]
```diff
@@ -548,9 +549,63 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
    @RequestMapping(value = "/mapping/hostnameip", method = RequestMethod.POST)
    public void createHostNameIPMapping(@RequestBody AssetIPMapping mapping) {
       BaseDocumentUtil.generateID(mapping);
+      String ip = mapping.getIp();
+      if(ip == null || ip.isEmpty() || !IPAddressUtil.isValidIp(ip)) {
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:553]: let's check if we already check empty in the validate method.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 13]
```diff
@@ -548,9 +549,63 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+   public void updateHostNameIPMapping(@RequestBody AssetIPMapping mapping) {
+      AssetIPMapping oldMapping = assetIPMappingRepository.findOne(mapping.getId());
+      if(oldMapping == null) {
+         throw new WormholeRequestException(HttpStatus.INTERNAL_SERVER_ERROR,"The dd is not exist : " + mapping.getId(),null);
+      }
+      String assetName = mapping.getAssetname();
+      if(assetName == null || !assetService.isAssetNameValidate(assetName)) {
+         throw new WormholeRequestException(HttpStatus.INTERNAL_SERVER_ERROR,"The Asset name is not exist : " + mapping.getAssetname(),null);
+      }
+      if(assetName.equals(oldMapping.getAssetname())) {
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:575]: let's move it before the server method.

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java (hunk 6 above)
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:578]: Done

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java (hunk 4 above)
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:567]: done

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java (hunk 1 above)
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java]: Done

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java (hunk 2 above)
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java]: Done

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java (hunk 3 above)
Comment (Pengpengwanga) [reply flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java]: Removed

File: flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java [hunk 14]
```diff
@@ -548,9 +549,63 @@ public ServerMapping getMappingById(@PathVariable("id") String id) {
...
+      }
       assetIPMappingRepository.save(mapping);
    }
 
+   @ResponseStatus(HttpStatus.OK)
+   @RequestMapping(value = "/mapping/hostnameip", method = RequestMethod.PUT)
+   public void updateHostNameIPMapping(@RequestBody AssetIPMapping mapping) {
+      AssetIPMapping oldMapping = assetIPMappingRepository.findOne(mapping.getId());
+      if(oldMapping == null) {
+         throw new WormholeRequestException(HttpStatus.INTERNAL_SERVER_ERROR,"The dd is not exist : " + mapping.getId(),null);
```
Comment (yixingjia) [inline flowgate-api/src/main/java/com/vmware/flowgate/controller/AssetController.java:569]: Let's fix it.  Cannot find an asset with id:

**Comment by Pengpengwanga** [comment]
Closed #398

**Pull request merged**